### PR TITLE
Create doc for gRPC iOS Network Transition Behavior

### PR DESCRIPTION
**Pull request by muxi**

### CODE REVIEW

File: src/objective-c/NetworkBehavior.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+Whenever there is a transition of network between these three categories, the
+previous channels is assumed to be broken. If there is an unfinished call, the
+call should return with status code `UNAVAILABLE`. All active gRPC channels
+will be destroyed so that any new call will trigger creation of new channel
+over new interface. In addition to that, when a TCP connection breaks, the
+corresponding channel should also be destroyed and calls be canceled with
+status code `UNAVAILABLE`.
+
+### Known issues
+There are several issues related to BSD sockets known to us that causes
```
Comment (srini100) [inline src/objective-c/NetworkBehavior.md]: cause

File: src/objective-c/NetworkBehavior.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+problems. 
+
+* TCP socket stalls but does not return error when network status transits from
+  Cellular to WiFi. This problem is workarounded by
```
Comment (srini100) [inline src/objective-c/NetworkBehavior.md]: resolved instead of workarounded

File: src/objective-c/NetworkBehavior.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+If you encounter these problems, the best solution is to switch to CFStream
+implementation which eliminates all of them.
+
+## gRPC iOS with CFStream
+gRPC iOS with CFStream implementation uses Apple's networking API to make
+connections. It resolves the issues above that is known to TCP sockets on iOS.
+Users are recommended to use this implementation rather than TCP socket
+implementation. With CFStream implementation, a channel is broken when the
+underlying stream detects an error or becomes closed. The behavior of streams
+in CFStream are not documented by Apple, but our experiments show that it is
```
Comment (srini100) [inline src/objective-c/NetworkBehavior.md]: change "are" to "is"

File: src/objective-c/NetworkBehavior.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+very similar to the model above, i.e. the streams error out when there is a
+network connetivity change. So users should expect channels to break when the
+network transits to another state and pending calls on those channels return
+with status code `UNAVAILABLE`.
```
Comment (srini100) [inline src/objective-c/NetworkBehavior.md]: Add a link to the CFStream guide.

File: src/objective-c/NetworkBehavior.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+be canceled with status code `UNAVAILABLE`.
+
+## Known issues
+gRPC currently uses BSD sockets for TCP connections. There are several issues
+related to BSD sockets known to us that causes problems. gRPC has a plan to
+switch to CFStream API for TCP connections which resolves some of these
+problems.
+
+* TCP socket stalls but does not return error when network status switches from
+  Cellular to WiFi. This problem is workarounded by
```
[the remark this answers is not in the document]
Comment (muxi) [reply src/objective-c/NetworkBehavior.md]: No. That direction is OK.

File: src/objective-c/NetworkBehavior.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+Whenever there is a switch of network connectivity between these three
+categories, the previous channels is assumed to be broken. If there is an
+unfinished call, the call should return with status code `UNAVAILABLE`. All
+active gRPC channels will be destroyed so that any new call will trigger
+creation of new channel over new interface. In addition to that, when a TCP
+connection breaks, the corresponding channel should also be destroyed and calls
+be canceled with status code `UNAVAILABLE`.
+
+## Known issues
+gRPC currently uses BSD sockets for TCP connections. There are several issues
```
[the remark this answers is not in the document]
Comment (muxi) [reply src/objective-c/NetworkBehavior.md]: BSD is subset of POSIX. On iOS it is BSD but both are correct.

**Comment by muxi** [comment]
Made some big changes in the doc. Please review again.

File: src/objective-c/NetworkTransitionBehavior.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
+
+# gRPC iOS Network Transition Behaviors
+Network connectivity on an iOS device may transition between cellular, WIFI, or
+no network connectivity. This document describes how these network changes
+should be handled by gRPC and current known issues.
```
[the remark this answers is not in the document]
Comment (muxi) [reply src/objective-c/NetworkTransitionBehavior.md]: Should be `should be`. TCP sockets are not behaving as `should be` as explained in later section.

File: src/objective-c/NetworkTransitionBehavior.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+destroyed. If there is an unfinished call, the call should return with status
+code `UNAVAILABLE`.
+
+`ConnectivityMonitor` is able to detect the scenario of the first issue above
+and actively destroy the channels. However, the second issue is not resolvable.
+To solve that issue the best solution is to switch to CFStream implementation
+which eliminates all of them.
+
+### gRPC iOS with CFStream
+gRPC iOS with CFStream implementation uses Apple's networking API to make
```
[the remark this answers is not in the document]
Comment (muxi) [reply src/objective-c/NetworkTransitionBehavior.md]: what release?

File: src/objective-c/NetworkTransitionBehavior.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+connections. It resolves the issues above that is known to TCP sockets on iOS.
+Users are recommended to use this implementation rather than TCP socket
+implementation. The detailed behavior of streams in CFStream is not documented
+by Apple, but our experiments show that it accords to the expected behaviors.
+With CFStream implementation, an event is always received when the underlying
+connection is no longer viable. For more detailed information and usages of
```
[the remark this answers is not in the document]
Comment (muxi) [reply src/objective-c/NetworkTransitionBehavior.md]: I think it is fine. `ConnectivityMonitor` is specific to TCP sockets as the structure of the doc suggests.

File: src/objective-c/NetworkTransitionBehavior.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (muxi) [reply src/objective-c/NetworkTransitionBehavior.md]: Oh I got it.

File: src/objective-c/NetworkTransitionBehavior.md [hunk 10]
```diff
@@ -27,34 +29,38 @@ The expected gRPC iOS channel and network transition behaviors are:
...
-the channel connection. Particularly, we found the following issues related to
-TCP sockets:
-* When a TCP sockets connection is established on cellular data and WiFi
+the channel connection. Particularly, we observed the following issues when
+using TCP sockets:
+* When a TCP socket connection is established on cellular data and WiFi
   becomes available, the TCP socket neither return an error event nor continue
   sending/receiving data on it, but still accepts write on it.
-* The TCP sockets does not report certain events that happens in the
+* A TCP sockets does not report certain events that happen in the
```
Comment (srini100) [inline src/objective-c/NetworkTransitionBehavior.md]: socket

File: src/objective-c/NetworkTransitionBehavior.md [hunk 11]
```diff
@@ -27,34 +29,38 @@ The expected gRPC iOS channel and network transition behaviors are:
...
   background. When a TCP connection breaks in the background for the reason
   like WiFi connects to another hotspot, the socket neither return an error nor
   continue sending/receiving data on it, but still accepts write on it.
+In both situations, the user will see the call hangs for an extended period of
```
Comment (srini100) [inline src/objective-c/NetworkTransitionBehavior.md]: hang

**Pull request merged**